libs/remote_storage/src/azure_blob.rs

```diff
@@ -335,3 +341,114 @@ impl RemoteStorage for AzureBlobStorage {
         Ok(())
     }
 }
+
+pin_project_lite::pin_project! {
+    #[project = NonSeekableStreamProj]
+    enum NonSeekableStream<S> {
+        Initial {
+            #[pin]
+            inner: tokio_util::compat::Compat<tokio_util::io::StreamReader<S, bytes::Bytes>>,
+            len: usize,
+            read_any: bool,
+        },
+        Cloned {
+            inner_was: std::marker::PhantomData<S>,
+            len_was: usize,
+        }
+    }
+}
+
+impl<S> NonSeekableStream<S>
+where
+    S: futures::stream::Stream<Item = std::io::Result<bytes::Bytes>> + Send + Sync + 'static,
+{
+    fn new(inner: S, len: usize) -> NonSeekableStream<S> {
+        use tokio_util::compat::TokioAsyncReadCompatExt;
+
+        let inner = tokio_util::io::StreamReader::new(inner).compat();
+        NonSeekableStream::Initial {
+            inner,
+            len,
+            read_any: false,
+        }
+    }
+}
+
+impl<S> std::fmt::Debug for NonSeekableStream<S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Initial { len, .. } => f.debug_struct("Initial").field("len", len).finish(),
+            Self::Cloned { len_was, .. } => f.debug_struct("Cloned").field("len", len_was).finish(),
+        }
+    }
+}
+
+impl<S> futures::io::AsyncRead for NonSeekableStream<S>
+where
+    S: futures::stream::Stream<Item = std::io::Result<bytes::Bytes>>,
+{
+    fn poll_read(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut [u8],
+    ) -> std::task::Poll<std::io::Result<usize>> {
+        match self.project() {
+            NonSeekableStreamProj::Initial {
+                inner, read_any, ..
+            } => {
+                *read_any = true;
+                inner.poll_read(cx, buf)
+            }
+            NonSeekableStreamProj::Cloned { .. } => std::task::Poll::Ready(Err(
+                std::io::Error::new(std::io::ErrorKind::Other, "cloned values cannot be read"),
+            )),
+        }
+    }
+}
+
+impl<S> Clone for NonSeekableStream<S> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Initial { len, .. } => Self::Cloned {
+                inner_was: std::marker::PhantomData,
+                len_was: *len,
+            },
+            Self::Cloned { inner_was, len_was } => Self::Cloned {
+                inner_was: *inner_was,
+                len_was: *len_was,
+            },
+        }
+    }
+}
```

Comment:
This does not work (yet) because they'll always do one clone, even without retrypolicy. I think there's a Mutex<Option<Inner>> way out of this but wasn't able to type one right away.